### PR TITLE
Extend expand_foreach test coverage (#296)

### LINKED ISSUE #296: Improve expand_foreach test coverage: foreach removal, edge cases, and Jinja filters
## Background

During review of PR #294, the following gaps in `src/manifest/expand_tests.rs` were identified. Addressing them is out of scope for that documentation-and-expansion PR.

## Work Required

File to edit: `src/manifest/expand_tests.rs`

1. In `expand_foreach_expands_sequence_values` (Test 1), after confirming `targets.len() == 2`, add an assertion for each expanded target that its object map does not contain the key `"foreach"`.

2. In `expand_foreach_applies_when_expression` (Test 2), after confirming the filtered count, add an assertion for each remaining target that its object map does not contain the key `"foreach"`.

3. In `expand_foreach_applies_action_when_expression` (Test 4), after confirming the filtered count, add an assertion for each remaining action that its object map does not contain the key `"foreach"`.

4. Add a test `expand_foreach_empty_foreach_produces_no_entries` that passes a target with `foreach: []` and asserts the expanded `targets` slice is empty.

5. Add a test `expand_foreach_non_object_entry_is_passed_through` that passes a `targets` array containing a bare string value (non-object) and asserts it survives unexpanded.

6. Add a test `expand_foreach_iteration_vars_do_not_overwrite_globals` that passes a target whose `vars` map already contains a key `"item"` with a value, and asserts the iteration-injected `item` value (from `foreach`) takes precedence for each expanded entry (or whichever precedence rule `expand_section` implements — assert the actual observed behaviour).

7. Add a test `expand_foreach_jinja_filter_in_name` that passes a target whose `name` uses a Jinja filter (e.g. `{{ item | upper }}`) with a `foreach` list of lowercase strings and asserts the expanded names are uppercased correctly.

Do not remove or modify any existing tests.

## AI Agent Prompt

~~~
File to edit: src/manifest/expand_tests.rs

1. In `expand_foreach_expands_sequence_values` (Test 1), after confirming
   `targets.len() == 2`, add an assertion for each expanded target that its
   object map does not contain the key `foreach`.

2. In `expand_foreach_applies_when_expression` (Test 2), after confirming
   the filtered count, add an assertion for each remaining target that its
   object map does not contain the key `foreach`.

3. In `expand_foreach_applies_action_when_expression` (Test 4), after
   confirming the filtered count, add an assertion for each remaining action
   that its object map does not contain the key `foreach`.

4. Add a test `expand_foreach_empty_foreach_produces_no_entries` that
   passes a target with `foreach: []` and asserts the expanded `targets`
   slice is empty.

5. Add a test `expand_foreach_non_object_entry_is_passed_through` that
   passes a `targets` array containing a bare string value (non-object) and
   asserts it survives unexpanded.

6. Add a test `expand_foreach_iteration_vars_do_not_overwrite_globals` that
   passes a target whose `vars` map already contains a key `item` with a
   value, and asserts the iteration-injected `item` value (from `foreach`)
   takes precedence for each expanded entry (or whichever precedence rule
   `expand_section` implements — assert the actual observed behaviour).

7. Add a test `expand_foreach_jinja_filter_in_name` that passes a target
   whose `name` uses a Jinja filter (e.g. `{{ item | upper }}`) with a
   `foreach` list of lowercase strings and asserts the expanded names are
   uppercased correctly.

Do not remove or modify any existing tests.
~~~

## Raised By

@leynos — identified during review of PR #294 (https://github.com/leynos/netsuke/pull/294).

## Rescope (2026-08-12)

The tests have since moved from `src/manifest/expand_tests.rs` into `src/manifest/expand_test_cases/`. Progress against the numbered items:

- Items 1–3 (foreach-key-removal assertions): **done** — `!map.contains_key("foreach")` assertions exist in `expand_test_cases/action_condition_cases.rs:27` and `condition_cases.rs:88`, and the proptest `foreach_key_is_removed_from_all_entries` (`property_cases.rs:97`, merged via #359) covers the invariant generally.
- Item 6 (iteration vars vs entry vars): **effectively covered** by `expand_foreach_when_item_overrides_entry_vars` (`condition_cases.rs:126`).
- Items 4, 5, and 7 (empty `foreach`, non-object passthrough, Jinja filter in `name`): **outstanding** — no such tests exist under `src/manifest/`. PR #358 is in flight to cover these.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Closes #296

Implements all seven items from the issue (the named tests now live in `src/manifest/expand_test_cases/condition_cases.rs` after the test-module split; `expand_foreach_expands_actions_sequence_values` already asserted `foreach` removal, so only the remaining gaps were added):

1. `expand_foreach_expands_sequence_values` — assert each expanded target lacks `foreach`.
2. `expand_foreach_applies_when_expression` — assert remaining targets lack `foreach`.
3. `expand_foreach_applies_action_when_expression` — assert remaining actions lack `foreach`.
4. New `expand_foreach_empty_foreach_produces_no_entries`.
5. New `expand_foreach_non_object_entry_is_passed_through`.
6. New `expand_foreach_iteration_vars_do_not_get_overwritten_by_entry_vars` — asserts the observed precedence: `inject_iteration_vars` inserts `item`/`index` after cloning, so the iteration value wins and unrelated vars survive.
7. New `expand_foreach_jinja_filter_in_name` — drives `manifest::from_str` (name rendering happens in `render_manifest`, not `expand_foreach`) and asserts uppercased names.

No existing test behaviour was removed or altered beyond the prescribed assertion additions; no production changes.

The pre-existing expand_foreach_preserves_object_key_order test remains unchanged in src/manifest/expand_test_cases/structure_cases.rs; it was moved solely to keep the condition test module within its enforced size limit.

Additional properties generate foreach sequences, source-key orderings, filters, and colliding entry variables. The Jinja name-rendering case remains an integration test.

## Validation

- `make check-fmt` — pass
- `make lint` — pass
- `make test` — pass (37 suites; expand suite now 36 tests)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Extend test coverage for foreach expansion in manifest targets and actions, including edge cases and name rendering behavior.

Tests:
- Add assertions that expanded targets and actions no longer contain foreach after expansion.
- Add tests for empty foreach producing no targets and for non-object target entries being passed through unchanged.
- Add tests ensuring iteration variables take precedence over entry vars while preserving unrelated vars.
- Add a test verifying Jinja filters in target names are correctly rendered to uppercase via the full manifest pipeline.

## References

- https://lody.ai/leynos/sessions/b08d2361-9870-4e96-b21d-836904d355ff

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[src/manifest/expand_test_cases/condition_cases.rs](https://github.com/leynos/netsuke/pull/358/files#diff-a529297f7a130f7640664629f5110f6c67edc4d1537c2978c50c214bb77720ca)

Comment on lines +255 to +263

~~~rust
    for action in actions {
        let map = action.as_object().context("action map")?;
        anyhow::ensure!(
            !map.contains_key("foreach"),
            "foreach should be removed from filtered actions"
        );
    }
    Ok(())
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69281/delta?repo-id=1160466&review-id=358&biomarker=Code+Duplication&filename=src%2Fmanifest%2Fexpand_test_cases%2Fcondition_cases.rs&method=)
The module contains 2 functions with similar structure: expand_foreach_applies_action_when_expression,expand_foreach_expands_actions_sequence_values

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: src/manifest/expand_test_cases/condition_cases.rs [hunk 1]
```diff
@@ -194,11 +195,105 @@ fn expand_foreach_applies_when_expression() -> Result<()> {
...
+    )?;
+    expand_foreach(&mut doc, &env)?;
+    let targets = targets(&doc)?;
+    anyhow::ensure!(targets.len() == 2, "expected both entries to survive");
+    anyhow::ensure!(
+        targets.first().and_then(ManifestValue::as_str) == Some("just-a-string"),
+        "bare string entry should pass through unexpanded: {:?}",
+        targets.first()
+    );
+    Ok(())
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/manifest/expand_test_cases/condition_cases.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The pass-through test uses only a bare string and a non-foreach object, so a no-op expand\_foreach would still return the expected two entries. | Add a foreach object beside the bare string and assert its expansion, item values, and ordering; then verify the bare string remains unchanged. |
| Linked Issues check | ⚠️ Warning | The PR implements all seven requested coverage items, but the summary indicates an existing object-key-order test was replaced. | Retain the existing object-key-order test unchanged, and add the Jinja filter test without replacing existing test coverage. |
| Testing (Property / Proof) | ⚠️ Warning | The test-only patch asserts foreach removal and key-order invariants, but uses fixed fixtures and adds no proptest coverage or substantive formal proof. | Add proptest cases for generated foreach sequences, entry values, map key orders, filtering, and variable precedence; retain integration tests for Jinja rendering. |

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by wafflecat-df12** [comment]
@coderabbitai review
